File: Games/Uzta/src/main/java/jokatu/game/games/uzta/game/ResourceDistributor.java
package jokatu.game.games.uzta.game;

import jokatu.game.games.uzta.event.GraphUpdatedEvent;
import jokatu.game.games.uzta.graph.LineSegment;
import jokatu.game.games.uzta.graph.Node;
import jokatu.game.games.uzta.graph.NodeType;
import jokatu.game.games.uzta.graph.UztaGraph;
import jokatu.game.games.uzta.player.UztaPlayer;
import ophelia.collections.bag.ModifiableIntegerBag;
import ophelia.event.observable.AbstractSynchronousObservable;
import ophelia.tuple.Pair;
import ophelia.util.MapUtils;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Stream;

import static java.util.Collections.emptyList;
import static java.util.Collections.emptySet;
import static java.util.stream.Collectors.*;
import static ophelia.collections.bag.BagCollectors.toBag;
import static ophelia.util.function.PredicateUtils.notNull;

class ResourceDistributor extends AbstractSynchronousObservable<GraphUpdatedEvent> {

	private final UztaGraph graph;
	private final Map<Node, Set<LineSegment>> nodeToOwningEdges;

	ResourceDistributor(UztaGraph graph) {
		this.graph = graph;

		nodeToOwningEdges = new HashMap<>();

		graph.getEdges().stream()
				.forEach(edge -> edge.forEach(node -> MapUtils.updateSetBasedMap(nodeToOwningEdges, node, edge)));
	}

	void distributeStartingResources() {
		Map<UztaPlayer, ModifiableIntegerBag<NodeType>> startingResources = graph.getEdges().stream()
				.filter(edge -> edge.getOwner() != null)
				.flatMap(this::getStartingResourcesForEdge)
				.collect(groupingBy(Pair::getLeft, mapping(Pair::getRight, toBag())));
		startingResources.forEach(UztaPlayer::giveResources);
	}

	private Stream<Pair<UztaPlayer, NodeType>> getStartingResourcesForEdge(@NotNull LineSegment edge) {
		UztaPlayer owner = edge.getOwner();
		assert owner != null;

		return edge.stream()
				.map(Node::getType)
				.map(type -> new Pair<>(owner, type));
	}

	void distributeResourcesForRoll(int roll) {

		Map<Boolean, List<Node>> nodesByRoll = graph.getNodes().stream()
				.collect(partitioningBy(node -> node.getValues().contains(roll)));

		List<Node> nodesForRoll = nodesByRoll.getOrDefault(true, emptyList());
		List<Node> nodesNotForRoll = nodesByRoll.getOrDefault(false, emptyList());

		nodesForRoll.forEach(node -> node.setHighlighted(true));
		nodesNotForRoll.forEach(node -> node.setHighlighted(false));

		fireEvent(new GraphUpdatedEvent(graph));

		Map<UztaPlayer, ModifiableIntegerBag<NodeType>> resources = nodesForRoll.stream()
				.flatMap(this::getResourcesForNode)
				.collect(groupingBy(Pair::getLeft, mapping(Pair::getRight, toBag())));
		resources.forEach(UztaPlayer::giveResources);
	}

	private Stream<Pair<UztaPlayer, NodeType>> getResourcesForNode(@NotNull Node node) {
		Set<LineSegment> edges = nodeToOwningEdges.getOrDefault(node, emptySet());
		return edges.stream()
				.map(LineSegment::getOwner)
				.filter(notNull())
				.map(owner -> new Pair<>(owner, node.getType()));
	}
}

File: Games/Uzta/src/main/kotlin/jokatu/game/games/uzta/game/ResourceDistributor.kt
package jokatu.game.games.uzta.game

import jokatu.game.games.uzta.event.GraphUpdatedEvent
import jokatu.game.games.uzta.graph.LineSegment
import jokatu.game.games.uzta.graph.Node
import jokatu.game.games.uzta.graph.NodeType
import jokatu.game.games.uzta.graph.UztaGraph
import jokatu.game.games.uzta.player.UztaPlayer
import ophelia.collections.bag.BagCollectors.toBag
import ophelia.event.observable.AbstractSynchronousObservable
import ophelia.tuple.Pair
import ophelia.util.function.PredicateUtils.notNull
import java.util.*
import java.util.Collections.emptyList
import java.util.Collections.emptySet
import java.util.stream.Collectors.*
import java.util.stream.Stream
import kotlin.collections.HashSet

typealias UserTappedNode = Pair<UztaPlayer, NodeType>

internal class ResourceDistributor(private val graph: UztaGraph) : AbstractSynchronousObservable<GraphUpdatedEvent>() {
	private val nodeToOwningEdges: MutableMap<Node, MutableSet<LineSegment>> = HashMap()

	init {
		graph.edges.stream()
				.forEach { edge ->
					edge.forEach { node ->
						nodeToOwningEdges.computeIfAbsent(node) { HashSet() }.add(edge)
					}
				}
	}

	fun distributeStartingResources() {
		val startingResources = graph.edges.stream()
				.filter { edge -> edge.owner != null }
				.flatMap { getStartingResourcesForEdge(it) }
				.collect(groupingBy({ n: UserTappedNode -> n.left }, mapping({ it.right }, toBag<NodeType>())))
		startingResources.forEach({ obj, givenResources -> obj.giveResources(givenResources) })
	}

	private fun getStartingResourcesForEdge(edge: LineSegment): Stream<Pair<UztaPlayer, NodeType>> {
		val owner = edge.owner!!

		return edge.stream()
				.map<NodeType> { it.type }
				.map { type -> Pair(owner, type) }
	}

	fun distributeResourcesForRoll(roll: Int) {

		val nodesByRoll = graph.nodes.stream()
				.collect(partitioningBy { node -> node.values.contains(roll) })

		val nodesForRoll = nodesByRoll.getOrDefault(true, emptyList())
		val nodesNotForRoll = nodesByRoll.getOrDefault(false, emptyList())

		nodesForRoll.forEach { node -> node.isHighlighted = true }
		nodesNotForRoll.forEach { node -> node.isHighlighted = false }

		fireEvent(GraphUpdatedEvent(graph))

		val resources = nodesForRoll.stream()
				.flatMap { getResourcesForNode(it) }
				.collect(groupingBy({ n: UserTappedNode -> n.left }, mapping({ it.right }, toBag<NodeType>())))
		resources.forEach { obj, givenResources -> obj.giveResources(givenResources) }
	}

	private fun getResourcesForNode(node: Node): Stream<UserTappedNode> {
		val edges = nodeToOwningEdges.getOrDefault(node, emptySet())
		return edges.stream()
				.map { it.owner }
				.filter(notNull())
				.map { owner -> Pair(owner!!, node.type!!) }
	}
}
